### ISSUE TITLE
delete error while delete all data from a big table

### ISSUE THREAD

**Issue by micklf** [greenplum-db/gpdb#1331]
```
db1=# select count(*) from lineitem;
   count   
-----------
 119994605
(1 row)
db1=# DELETE FROM lineitem ;
ERROR:  lock 4551 is not held (lwlock.c:728)  (seg17 192.168.181.214:40008 pid=13808) (cdbdisp.c:223)
```

found infomations from the log:
```
2016-11-16 11:08:40.961504 CST,"xxxx","db1",p16126,th-877041600,"192.168.181.211","52387",2016-11-16 10:23:47 CST,222603,con3865,cmd96,seg-1,,dx4620,x222603,sx1,"LOG","00000","statement: delete from lineitem ;",,,,,,"delete from lineitem ;",0,,"postgres.c",1610,
2016-11-16 11:08:41.040441 CST,"lxr1","db1",p16126,th-877041600,"192.168.181.211","52387",2016-11-16 10:23:47 CST,222603,con3865,cmd97,seg-1,,dx4620,x222603,sx1,"ERROR","XX000","lock 2413 is not held (lwlock.c:728)  (seg16 192.168.181.213:40008 pid=6798) (cdbdisp.c:223)",,,,,,"delete from lineitem ;",0,,"cdbdisp.c",223,"Stack trace:
1    0x905148 postgres errstart + 0x298
2    0x9831d5 postgres cdbdisp_finishCommand + 0x1f5
3    0x65f62f postgres mppExecutorFinishup + 0x19f
4    0x64723c postgres ExecutorEnd + 0xfc
5    0x7ec656 postgres <symbol not found> + 0x7ec656
6    0x7eca92 postgres <symbol not found> + 0x7eca92
7    0x7edb3e postgres PortalRun + 0x2ce
8    0x7e808b postgres <symbol not found> + 0x7e808b
9    0x7ea6c4 postgres PostgresMain + 0x1ce4
10   0x788ec4 postgres <symbol not found> + 0x788ec4
11   0x78bb59 postgres PostmasterMain + 0x789
12   0x4893eb postgres main + 0x3bb
13   0x7f69ca63cb15 libc.so.6 __libc_start_main + 0xf5
14   0x489509 postgres <symbol not found> + 0x489509
"
```

In big table whith many lines, say 100 million lines, Can not use `delete from tablexxxx` ?

**Comment by hlinnaka** [comment]
Hmm. That stack trace isn't very helpful, unfortunately, as it's from the master node, rather than the segment where the error happened. Can you look into seg17's log, to see if there's a more useful stack trace?

**Comment by vraghavan78** [comment]
@asimrp @ashwinstar @yydzero can one of you look into this?

**Comment by micklf** [comment]
@hlinnaka I tried again, and got the logs below on the segment, the error one got from master's log.
```
2016-11-18 16:40:52.954479 CST,"gpadmin","xxxxx",p20416,th-1979602880,"192.168.181.211","49291",2016-11-18 16:40:35 CST,111805,con19,cmd6,seg10,,dx4,x111805,sx1,"ERROR","XX000","lock 1135 is not held (lwlock.c:728)",,,,,,"delete from lineitem;",0,,"lwlock.c",728,"Stack trace:
1    0x905148 postgres errstart (??:?)
2    0x906f0b postgres elog_finish (??:?)
3    0x7d90c5 postgres LWLockRelease (??:?)
4    0x7be253 postgres <symbol not found> (bufmgr.c:?)
5    0x7be729 postgres ReadBufferWithStrategy (??:?)
6    0x4ad906 postgres <symbol not found> (heapam.c:?)
7    0x4ae25c postgres <symbol not found> (heapam.c:?)
8    0x4b308d postgres heap_getnextx (??:?)
9    0x6962e4 postgres HeapScanNext (??:?)
10   0x659792 postgres ExecScan (??:?)
11   0x680ddc postgres ExecTableScan (??:?)
12   0x64d30f postgres ExecProcNode (??:?)
13   0x64ab5a postgres <symbol not found> (execMain.c:?)
14   0x64b13d postgres ExecutorRun (??:?)
15   0x7ec633 postgres <symbol not found> (pquery.c:?)
16   0x7eca92 postgres <symbol not found> (pquery.c:?)
17   0x7edb3e postgres PortalRun (??:?)
18   0x7e557d postgres <symbol not found> (postgres.c:?)
19   0x7ea9f7 postgres PostgresMain (??:?)
20   0x788ec4 postgres <symbol not found> (postmaster.c:?)
21   0x78bb59 postgres PostmasterMain (??:?)
22   0x4893eb postgres main (??:?)
23   0x7f1988ac0b15 libc.so.6 __libc_start_main (??:0)
24   0x489509 postgres <symbol not found> (??:?)
"
2016-11-18 16:40:52.972037 CST,"gpadmin","xxxx",p20416,th-1979602880,"192.168.181.211","49291",2016-11-18 16:40:35 CST,0,con19,,seg10,,dx4,,sx1,"DEBUG1","00000","Message type T received by from libpq, len = 78",,,,,,,0,,"postgres.c",5146,
2016-11-18 16:41:10.973223 CST,"gpadmin","adb1",p20416,th-1980008704,"192.168.181.211","49291",2016-11-18 16:40:35 CST,0,con19,,seg10,,,,,"LOG","00000","udp-ic: rx-thread shutting down",,,,,,,0,,,,
2016-11-18 16:41:10.998518 CST,"gpadmin","adb1",p20416,th-1979602880,"192.168.181.211","49291",2016-11-18 16:40:35 CST,0,con19,,seg10,,,,,"DEBUG1","00000","SessionState_Release: pinCount: 0, activeProcessCount: 0",,,,,,,0,,"session_state.c",154,"Stack trace:
1    0x905148 postgres errstart (??:?)
2    0x96dea3 postgres SessionState_Shutdown (??:?)
3    0x9167f1 postgres <symbol not found> (postinit.c:?)
4    0x7ca195 postgres proc_exit_prepare (??:?)
5    0x7ca239 postgres proc_exit (??:?)
6    0x7ea191 postgres PostgresMain (??:?)
7    0x788ec4 postgres <symbol not found> (postmaster.c:?)
8    0x78bb59 postgres PostmasterMain (??:?)
9    0x4893eb postgres main (??:?)
10   0x7f1988ac0b15 libc.so.6 __libc_start_main (??:0)
```

**Comment by hlinnaka** [comment]
Thanks!

Which version did you build this from? This looks a lot like the bug that was fixed in commit 3d3d3cc2, but that was fixed on Sep 1st already. Is it possible that your checkout is older than that?

If that's not it, any chance you could re-build with --enable-debug, to get accurate line number information in there?

**Comment by micklf** [comment]
@hlinnaka I got the source code at about August. I'll try to review commit  3d3d3cc, open the debug mode to try to solve this.

If it is caused by 3d3d3cc, i'll close this issue ASAP.

**Comment by micklf** [comment]
@hlinnaka yes, i merged the commit of  3d3d3cc, which could fix the bug.
